config/experimental/config/config.go

```diff
@@ -0,0 +1,61 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"go.opentelemetry.io/collector/config"
+)
+
+// SourceSettings defines common settings of a Source configuration.
+// Specific config sources can embed this struct and extend it with more fields if needed.
+// When embedded it must be with `mapstructure:",squash"` tag.
+type SourceSettings struct {
+	config.ComponentID `mapstructure:"-"`
+}
+
+// ID returns the ID of the component that this configuration belongs to.
+func (s *SourceSettings) ID() config.ComponentID {
+	return s.ComponentID
+}
+
+// SetIDName updates the name part of the ID for the component that this configuration belongs to.
+func (s *SourceSettings) SetIDName(idName string) {
+	s.ComponentID = config.NewIDWithName(s.ComponentID.Type(), idName)
+}
+
+// NewSourceSettings return a new config.SourceSettings struct with the given ComponentID.
+func NewSourceSettings(id config.ComponentID) SourceSettings {
+	return SourceSettings{id}
+}
+
+// Source is the configuration of a config source. Specific config sources must implement this
+// interface and will typically embed SourceSettings struct or a struct that extends it.
+type Source interface {
+	// TODO: While config sources are experimental and not in the config package they can't
+	//	 reference the private interfaces config.identifiable and config.validatable.
+	//   Defining the required methods of the interfaces temporarily here.
+
+	// From config.identifiable:
+
+	// ID returns the ID of the component that this configuration belongs to.
+	ID() config.ComponentID
+	// SetIDName updates the name part of the ID for the component that this configuration belongs to.
+	SetIDName(idName string)
```

Comment:
What is there are setter? What's the use case to rename the ID in runtime?

Reply:
This is similar to the other components in the collector that you can have named exporters, receivers, and processors. This can be useful in the case of config sources that have some configuration settings, for example:

```yaml
config_sources:
    # Assuming that the environment variables VAULT_ADDR and VAULT_TOKEN are the defined
    # and the different secrets are on the same server but at different paths.
    vault/kv:
      endpoint: $VAULT_ADDR
      path: secret/data/kv
      auth:
        token: $VAULT_TOKEN
    vault/db:
      endpoint: $VAULT_ADDR
      path: database/creds/collector_role
      auth:
        token: $VAULT_TOKEN

# Both Vault config sources can be used via their full name. Hypothetical example:
components:
  component_using_vault_kv:
    # Example showing K/V V2, see note below about the '.' usage.
    username: $vault/kv:data.user
    password: $vault/kv:data.password

  component_using_vault_db:
    username: $vault/db:username
    password: $vault/db:password
```